lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py

```diff
@@ -0,0 +1,110 @@
+from typing import TYPE_CHECKING
+import h5py
+import numpy as np
+import zarr
+
+from ..LindiH5pyDataset import LindiH5pyDataset
+from ..LindiH5pyReference import LindiH5pyReference
+
+if TYPE_CHECKING:
+    from ..LindiH5pyGroup import LindiH5pyGroup  # pragma: no cover
+
+from ...conversion.create_zarr_dataset_from_h5_data import create_zarr_dataset_from_h5_data
+
+
+class LindiH5pyGroupWrite:
+    def __init__(self, p: 'LindiH5pyGroup'):
+        self.p = p
+
+    def create_group(self, name, track_order=None):
+        from ..LindiH5pyGroup import LindiH5pyGroup  # avoid circular import
+        if track_order is not None:
+            raise Exception("track_order is not supported (I don't know what it is)")
+        if isinstance(self.p._group_object, h5py.Group):
+            return LindiH5pyGroup(
+                self.p._group_object.create_group(name), self.p._file
+            )
+        elif isinstance(self.p._group_object, zarr.Group):
+            return LindiH5pyGroup(
+                self.p._group_object.create_group(name), self.p._file
+            )
+        else:
+            raise Exception(f'Unexpected group object type: {type(self.p._group_object)}')
+
+    def require_group(self, name):
+        if name in self.p:
+            ret = self.p[name]
+            if not isinstance(ret, LindiH5pyGroup):
+                raise Exception(f'Expected a group at {name} but got {type(ret)}')
+            return ret
+        return self.create_group(name)
+
+    def create_dataset(self, name, shape=None, dtype=None, data=None, **kwds):
```

Comment:
There are a few places in HDMF that call `group.require_dataset` when writing datasets containing references, e.g., https://github.com/hdmf-dev/hdmf/blob/244d17a28ed436849b1973a3aaac8522d0ea922b/src/hdmf/backends/hdf5/h5tools.py#L1185

I do not know why `require_dataset` is called instead of `create_dataset` but that's what it is right now...

It looks like the source code for `Group.require_dataset` returns an `h5py.Dataset` if the dataset exists already. https://github.com/h5py/h5py/blob/066cad17312a592b26f59f91ef383dbde29bd142/h5py/_hl/group.py#L281-L282

Should we return an `LindiH5pyDataset` instead?

Reply:
Thanks. It was using the `require_dataset()` from the base class, which is not what we want. I have now implemented require_dataset on the LindiH5pyGroup, added a test, and pushed the update.

Reply:
I don't see `require_dataset` anywhere. Where did you add it?

Reply:
Never mind - page did not refresh